premium/usage.go

```diff
@@ -543,54 +560,79 @@ func (u *BatchUpdater) reportUsageToAWSMarketplace(ctx context.Context, rows uin
 		UsageQuantity:    aws.Int32(int32(rows)),
 	})
 	if err != nil {
-		return fmt.Errorf("failed to update usage with : %w", err)
+		return fmt.Errorf("failed to update usage: %w", err)
 	}
 	return nil
 }
 
-func (u *BatchUpdater) updateUsageWithRetryAndBackoff(ctx context.Context, rows uint32, tables []cqapi.UsageIncreaseTablesInner) error {
+func (u *BatchUpdater) updateMarketplaceUsageWithRetryAndBackoff(ctx context.Context, rows uint32) error {
+	var lastErr error
 	for retry := 0; retry < u.maxRetries; retry++ {
-		u.logger.Debug().Str("url", u.url).Int("try", retry).Int("max_retries", u.maxRetries).Uint32("rows", rows).Msg("updating usage")
+		u.logger.Debug().Int("try", retry).Int("max_retries", u.maxRetries).Uint32("rows", rows).Msg("updating usage")
 		queryStartTime := time.Now()
 
-		// If the AWS Marketplace client is set, use it to track usage
-		if u.awsMarketplaceClient != nil {
-			return u.reportUsageToAWSMarketplace(ctx, rows)
-		}
-		payload := cqapi.IncreaseTeamPluginUsageJSONRequestBody{
-			RequestId:  uuid.New(),
-			PluginTeam: u.pluginMeta.Team,
-			PluginKind: u.pluginMeta.Kind,
-			PluginName: u.pluginMeta.Name,
-			Rows:       int(rows),
+		lastErr = u.reportUsageToAWSMarketplace(ctx, rows)
+		if lastErr == nil {
+			u.logger.Debug().Int("try", retry).Uint32("rows", rows).Msg("usage updated")
+			return nil
 		}
 
-		if len(tables) > 0 {
-			payload.Tables = &tables
+		var de *types.DuplicateRequestException
+		if errors.As(lastErr, &de) {
+			jitter := time.Duration(rand.Intn(1000)) * time.Millisecond
```

Comment:
I've seen this already used in this code. Wouldn't it make more sense for the jitter max to be proportional to the backoff, say 5% of it, instead of hardcoded to 1 second?

Reply:
Here the backoff is constant at 1 sec, due to how we want the data to be registered to the next second (to prevent `DuplicateRequestException`, which happens because presumably we're running multiple instances and everybody's reporting their usage at the same second)

Because of this, the jitter max value can be increased to 2sec or more (as we only care about 1-second blocks...), but having it lower also makes it sense that it would finish quicker.

Reply:
The other part where this code lives is for the "regular" retries (not DuplicateRequestException) which is implemented on top of the (presumed already existing) AWS SDK backoff mechanisms. We could remove that cases and set up the AWS SDK backoff params instead, but I'm not an expert on that. This also protects against similar errors (if any that we've potentially missed, or is undocumented) that is not covered by the `errors.As`.